src/xdp-sealed-fd.c

```diff
@@ -0,0 +1,226 @@
+  int saved_errno = -1;
+  g_autoptr(XdpSealedFd) sealed_fd = NULL;
+  char *shm;
+  g_autoptr(GOutputStream) stream = NULL;
+
+  g_return_val_if_fail (bytes != NULL, NULL);
+
+  if ((fd = memfd_create ("sealed-fd-from-bytes", MFD_ALLOW_SEALING)) == -1)
+    {
+      int saved_errno;
+
+      saved_errno = errno;
+      g_set_error (error,
+                   G_IO_ERROR,
+                   g_io_error_from_errno (saved_errno),
+                   "memfd_create: %s", g_strerror (saved_errno));
+      return NULL;
+    }
+
+  bytes_data = g_bytes_get_data (bytes, &bytes_len);
+
+  if ((shm = (char *) mmap (NULL, bytes_len, PROT_WRITE, MAP_SHARED, fd, 0)) == (void *) -1)
+    {
+      int saved_errno;
+
+      saved_errno = errno;
+      g_set_error (error,
+                   G_IO_ERROR,
+                   g_io_error_from_errno (saved_errno),
+                   "mmap: %s", g_strerror (saved_errno));
+      return NULL;
+    }
+
+  if (ftruncate (fd, bytes_len) == -1)
+    {
+      int saved_errno;
+
+      saved_errno = errno;
+      g_set_error (error,
+                   G_IO_ERROR,
+                   g_io_error_from_errno (saved_errno),
+                   "ftruncate: %s", g_strerror (saved_errno));
+      return NULL;
+    }
+
+  memcpy (shm, bytes_data, bytes_len);
+
+  if (munmap (shm, bytes_len) == -1)
+    {
+      int saved_errno;
+
+      saved_errno = errno;
+      g_set_error (error,
+                   G_IO_ERROR,
+                   g_io_error_from_errno (saved_errno),
+                   "munmap: %s", g_strerror (saved_errno));
+      return NULL;
+    }
+
+  if (fcntl (fd, F_ADD_SEALS, required_seals) == -1)
+    {
+      saved_errno = errno;
+      g_set_error (error,
+                   G_IO_ERROR,
+                   g_io_error_from_errno (saved_errno),
+                   "fcntl F_ADD_SEALS: %s", g_strerror (saved_errno));
+      return NULL;
+    }
+
+  sealed_fd = g_object_new (XDP_TYPE_SEALED_FD, NULL);
+  sealed_fd->fd = xdp_steal_fd (&fd);
+
+  return g_steal_pointer (&sealed_fd);
+}
+
+XdpSealedFd *
+xdp_sealed_fd_new_from_handle (GVariant     *handle,
+                               GUnixFDList  *fd_list,
+                               GError      **error)
+{
+  int fd_id;
+  xdp_autofd int fd = -1;
+
+  g_return_val_if_fail (g_variant_is_of_type (handle, G_VARIANT_TYPE_HANDLE), NULL);
+  g_return_val_if_fail (G_IS_UNIX_FD_LIST (fd_list), NULL);
+  g_return_val_if_fail (g_unix_fd_list_get_length (fd_list) > 0, NULL);
+
+  fd_id = g_variant_get_handle (handle);
+  if ((fd = g_unix_fd_list_get (fd_list, fd_id, error)) == -1)
+    return NULL;
+
+  return xdp_sealed_fd_new_take_memfd (xdp_steal_fd (&fd), error);
+}
+
+int
+xdp_sealed_fd_get_fd (XdpSealedFd *sealed_fd)
+{
+  g_return_val_if_fail (XDP_IS_SEALED_FD (sealed_fd), -1);
+
+  return sealed_fd->fd;
+}
+
+int
+xdp_sealed_fd_dup_fd (XdpSealedFd *sealed_fd)
+{
+  g_return_val_if_fail (XDP_IS_SEALED_FD (sealed_fd), -1);
+
+  return dup (sealed_fd->fd);
```

Comment:
dup can fail

Reply:
Right, I didn't think about it, but actually the only reason it can fail is that the we have to many fds open. https://www.man7.org/linux/man-pages/man2/dup.2.html

Reply:
@swick how should I handle a failure? just return an error and then log it? Do you have any examples in glib

Reply:
> just return an error and then log it?

Probably. There isn't a whole lot you can do in those situations.